topics/go/algorithms/data/heap/heap.go

```diff
@@ -0,0 +1,140 @@
+// Package heap implements a heap interface, and add ability to store values of type Data in the minimum Heap.
+package heap
+
+import (
+	"container/heap"
+	"errors"
+)
+
+// ErrEmptyHeap returned if heap is empty
+var ErrEmptyHeap = errors.New("heap is empty")
+
+// Data represents what is being stored on the heap.
+type Data struct {
+	Value int
+	Index int
+}
+
+// Heap represents a list of data.
+type Heap struct {
+	size     int
+	capacity int
+	dataHeap *dataHeap
+}
+
+// New returns a new Heap with given capacity in which we will store our data.
+func New(cap int) (*Heap, error) {
+
+	if cap <= 0 {
+		return nil, errors.New("invalid capacity")
+	}
+
+	h := Heap{
+		size:     0,
+		capacity: cap,
+		dataHeap: &dataHeap{},
+	}
+
+	return &h, nil
+}
+
+// Store is storing the element of type Data in the heap.
+func (h *Heap) Store(data *Data) error {
+
+	// If we try to add data to full Heap, we will return an error.
+	if h.size >= h.capacity {
+		return errors.New("getting out of heap capacity")
+	}
+
+	// The complexity of the heap.Push operation is O(log n) where n = h.dataHeap.Len().
+	// The element which we will be store in the heap will be added closely to the end of the heap and after what it
+	// will be sift up, until it not get in the right position in the heap according it's value.
+	heap.Push(h.dataHeap, data)
+	h.size++
+	return nil
+}
+
+// Extract returns the minimum element of type Data from the heap or error error heap is empty.
+// Since we have minimum heap implementation. The size of the heap will be decreased.
+func (h *Heap) Extract() (*Data, error) {
+
+	// Check that heap is not empty
+	if len(*h.dataHeap) == 0 {
+		return nil, ErrEmptyHeap
+	}
+
+	// heap.Pop return the element from the heap. the type of element is empty interface.
+	element := heap.Pop(h.dataHeap)
+	h.size--
+
+	// Since the returned value we got has an empty interface type, let's convert it to our Data type.
```

Comment:
// Perform a type assertion to get a copy of the Data pointer stored inside of element.